### PR TITLE
[wasm] Fix circular dependency in WasmAppBuilder

### LINKED ISSUE #58816: Source-build encounters circular PublishBuilder dependency in WasmAppBuilder
This issue was discovered while updating source-build to the latest 6.0 commit shas - https://github.com/dotnet/installer/pull/11710/files

/repos/tarball1/src/runtime.c300b096419523024f2b807ec9db3c2d91df0298/artifacts/source-build/self/src/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj(33,5): error MSB4006: There is a circular dependency in the target dependency graph involving target "PublishBuilder".

Binlog - [runtime.circularbuild.sourcebuild.zip](https://github.com/dotnet/runtime/files/7130269/runtime.circularbuild.sourcebuild.zip)

This is what the call graph looks like:
MSBuild WasmAppBuilder net6.0;net472
	WasmAppBuilder net6.0 -> Build
		WasmAppBuilder net472 -> publish
			MSBuild
			    Parameters
			        Projects = WasmAppBuilder.csproj
			        Targets = Publish
			        Properties = TargetFramework=net472
			    Global Properties:
			        TargetFramework = net472

This appears to be caused by this target in WasmAppBuilder.csproj - https://github.com/dotnet/runtime/blob/release/6.0/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj#L26

```
  <Target Name="PublishBuilder"
          AfterTargets="Build">

    <!-- needed for publishing with multi-targeting. We are publishing essentially to get the SR.MetadataLoadContext.dll :/ -->
    <ItemGroup>
      <_PublishFramework Include="$(TargetFrameworks)" />
    </ItemGroup>
    <MSBuild Projects="$(MSBuildProjectFile)" Targets="Publish" Properties="TargetFramework=%(_PublishFramework.Identity)" />
  </Target>
```

### PR DESCRIPTION
**Pull request by radical**
Fixes https://github.com/dotnet/runtime/issues/58816

### CODE REVIEW

**Comment by MichaelSimons** [comment]
A friendly reminder to port to release/6.0.

I'm not knowledgeable enough in this area to provide a code review.

**Review by lewing — APPROVED** [review]

**Comment by radical** [comment]
Windows test failures are https://github.com/dotnet/runtime/issues/58927 .

**Comment by radical** [comment]
`iOSSimulator` build is failing with multiple errors in copying files with `because it is being used by another process`:

```sh
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/.packages/system.threading.tasks.dataflow/4.9.0/lib/netstandard2.0/System.Threading.Tasks.Dataflow.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Threading.Tasks.Dataflow.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Threading.Tasks.Dataflow.dll' because it is being used by another process. 
/Users/runner/work/1/s/.dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: Could not copy "/Users/runner/work/1/s/.packages/system.reflection.metadataloadcontext/4.7.1/lib/netstandard2.0/System.Reflection.MetadataLoadContext.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Reflection.MetadataLoadContext.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Reflection.MetadataLoadContext.dll' because it is being used by another process.  [/Users/runner/work/1/s/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj]
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/.packages/system.reflection.metadataloadcontext/4.7.1/lib/netstandard2.0/System.Reflection.MetadataLoadContext.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Reflection.MetadataLoadContext.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Reflection.MetadataLoadContext.dll' because it is being used by another process. 
/Users/runner/work/1/s/.dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: Could not copy "/Users/runner/work/1/s/.packages/system.reflection.metadata/1.8.0/lib/netstandard2.0/System.Reflection.Metadata.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Reflection.Metadata.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Reflection.Metadata.dll' because it is being used by another process.  [/Users/runner/work/1/s/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj]
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/.packages/system.reflection.metadata/1.8.0/lib/netstandard2.0/System.Reflection.Metadata.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Reflection.Metadata.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Reflection.Metadata.dll' because it is being used by another process. 
  WasmBuildTasks -> /Users/runner/work/1/s/artifacts/bin/WasmBuildTasks/Debug/net6.0/WasmBuildTasks.dll
/Users/runner/work/1/s/.dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: Could not copy "/Users/runner/work/1/s/.packages/system.security.principal.windows/4.7.0/lib/net461/System.Security.Principal.Windows.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Security.Principal.Windows.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Security.Principal.Windows.dll' because it is being used by another process.  [/Users/runner/work/1/s/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj]
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/.packages/system.security.principal.windows/4.7.0/lib/net461/System.Security.Principal.Windows.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Security.Principal.Windows.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Security.Principal.Windows.dll' because it is being used by another process. 
/Users/runner/work/1/s/.dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: Could not copy "/Users/runner/work/1/s/.packages/system.security.principal.windows/4.7.0/lib/net461/System.Security.Principal.Windows.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Security.Principal.Windows.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Security.Principal.Windows.dll' because it is being used by another process.  [/Users/runner/work/1/s/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj]
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/.packages/system.security.principal.windows/4.7.0/lib/net461/System.Security.Principal.Windows.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Security.Principal.Windows.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Security.Principal.Windows.dll' because it is being used by another process. 
/Users/runner/work/1/s/.dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: Could not copy "/Users/runner/work/1/s/.packages/system.text.json/4.7.0/lib/net461/System.Text.Json.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Text.Json.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Text.Json.dll' because it is being used by another process.  [/Users/runner/work/1/s/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj]
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/.packages/system.text.json/4.7.0/lib/net461/System.Text.Json.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Text.Json.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Text.Json.dll' because it is being used by another process. 
/Users/runner/work/1/s/.dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: Could not copy "/Users/runner/work/1/s/.packages/system.text.json/4.7.0/lib/net461/System.Text.Json.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Text.Json.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Text.Json.dll' because it is being used by another process.  [/Users/runner/work/1/s/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj]
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(4823,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/.packages/system.text.json/4.7.0/lib/net461/System.Text.Json.dll" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Text.Json.dll". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/System.Text.Json.dll' because it is being used by another process. 
  WasmBuildTasks -> /Users/runner/work/1/s/artifacts/bin/WasmBuildTasks/Debug/net6.0/publish/
  WasmAppBuilder -> /Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll
/Users/runner/work/1/s/.dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(5174,5): error MSB3026: Could not copy "/Users/runner/work/1/s/artifacts/obj/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config' because it is being used by another process.  [/Users/runner/work/1/s/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj]
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(5174,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/artifacts/obj/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config' because it is being used by another process. 
/Users/runner/work/1/s/.dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(5174,5): error MSB3026: Could not copy "/Users/runner/work/1/s/artifacts/obj/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config' because it is being used by another process.  [/Users/runner/work/1/s/src/tasks/WasmAppBuilder/WasmAppBuilder.csproj]
##[error].dotnet/sdk/6.0.100-rc.1.21430.12/Microsoft.Common.CurrentVersion.targets(5174,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "/Users/runner/work/1/s/artifacts/obj/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config" to "/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config". Beginning retry 1 in 1000ms. The process cannot access the file '/Users/runner/work/1/s/artifacts/bin/WasmAppBuilder/Debug/net472/WasmAppBuilder.dll.config' because it is being used by another process. 
```

Are these just intermittent errors, or is this PR breaking something?
@steveisok any idea?

**Comment by steveisok** [comment]
First I've seen something like this 🤷‍♂️

**Comment by radical** [comment]
The test failures are on Windows11 - #58927, unrelated to this PR.

File: src/tasks/WasmAppBuilder/WasmAppBuilder.csproj [hunk 1]
```diff
@@ -24,13 +24,14 @@
...
-          AfterTargets="Build">
+          AfterTargets="Build"
+          Condition="'$(_RunningForPublishBuilder)' != 'true'">
 
     <!-- needed for publishing with multi-targeting. We are publishing essentially to get the SR.MetadataLoadContext.dll :/ -->
     <ItemGroup>
       <_PublishFramework Include="$(TargetFrameworks)" />
     </ItemGroup>
-    <MSBuild Projects="$(MSBuildProjectFile)" Targets="Publish" Properties="TargetFramework=%(_PublishFramework.Identity)" />
+    <MSBuild Projects="$(MSBuildProjectFile)" Targets="Publish" Properties="TargetFramework=%(_PublishFramework.Identity);_RunningForPublishBuilder=true" />
```
Comment (ericstj) [inline src/tasks/WasmAppBuilder/WasmAppBuilder.csproj:34]: Global properties added to MSBuild invocation. This is a recipe for a build race condition. The global property defeats the MSBuild results cache and also can cause concurrent builds to the same path if the global property doesn’t change output directories.

**Pull request merged**